File: Python/forward_backward_optimisation.py
# -*- coding: utf-8 -*-
"""
Basic Pursuit with Forward-Backward

Created on Tue Mar 30 11:19:42 2021

@author: caroline-pinte
"""

import numpy as np

#====================================================================
# Utilities
#====================================================================

def F(x,l) :
    return l * np.linalg.norm(x,1)

#=========

# Not tested
def G(A,x,y) :
    return 1/2 * np.linalg.norm(y-A*x,2)**2

def G3(A,x,y) :
    G = 0
    for i in range(0,np.shape(A)[0]) :
        G = G + (y[i] - np.trace( np.matmul(np.squeeze(A[i,:,:]).T,x) ))**2
    return G

#=========

# Not tested
def GradG(A,x,y) :
    return np.matmul(A.T , (np.matmul(A,x)-y))

def GradG3(A,x,y) :
    gradG = np.zeros(np.shape(x))
    for i in range(0,np.shape(A)[0]) :
        gradG = gradG + np.squeeze(A[i,:,:])*(np.trace( np.matmul(np.squeeze(A[i,:,:]).T,x) ) - y[i])
    return gradG

#=========

def prox_L21_1(x,l,mu) :
    prox = np.zeros((np.shape(x)[0],np.shape(x)[1]))
    for p in range(0,np.shape(x)[0]) :
        for k in range(0,np.shape(x)[1]) :
            div_ = 0
            for kk in range(0,np.shape(x)[1]) :
                div_ = div_ + max(abs(x[p,kk])-mu , 0)**2
            div_ = np.sqrt(div_)
            if (div_ == 0) :
                frac = 0
            else :
                frac = l/div_
            prox[p,k] = np.sign(x[p,k]) * max(abs(x[p,k])-mu , 0) * max((1 - frac) , 0)
    return prox
 
def ProxF(x,tau,l,rho) :
    return prox_L21_1(x, l*tau, rho*tau)

#=========

def perform_fb(x, L, method, niter, A, y, l, rho) :
    eps = 0.0003 # or 0.0004 if too slow
    
    t = 1
    Y = x.copy()
    
    R = []
    for i in range(0,niter) :
        #print(i)
        f = F(x,l)
        if (len(np.shape(A)) == 3) :
            g = G3(A,x,y)
        else :
            g = G(A,x,y)
        R.append( f+g ) 
        
        if (method == 'fista') :
            if (len(np.shape(A)) == 3) :
                grad = GradG3(A,Y,y)
            else :
                grad = GradG(A,Y,y)
            xnew = ProxF( Y - 1/L*grad, 1/L, l, rho );
            tnew = (1+np.sqrt(1+4*t**2)) / 2
            Y = xnew + (t-1)/(tnew)*(xnew-x)
            x = xnew
            t = tnew
            # stopping criteria
            if (i>1) and ((R[i-1]-R[i])/R[i] < eps) :
                 break

        else :
            print('Error in perform_fb : name isnt fista')
            return
    return x

#====================================================================
# forward_backward_optimisation
#====================================================================
def forward_backward_optimisation(A, y, l, method_i=1, rho=1500) :
    '''
    Compute the optimal activation pattern : alpha.

            Parameters:
                    A (array): slice of design matrix for learning.
                    y (array): slice of NF variable for learning.
                    l (int): lambda to be tested.
                    method_i (int): regularisation function, 1 corresponds to fistaL1.
                    rho (int): regularisation parameter rho.
                    
            Returns:
                    alpha (array): optimal activation pattern
    '''
    
    n = np.shape(A)[1]*np.shape(A)[2]
    
    # Lipschitz constant.
    if (len(np.shape(A)) == 2) :
        L = np.linalg.norm(A,2)
    elif (len(np.shape(A)) == 3) :
        # size AA: d2 x d1*d3
        test = np.squeeze(A[0,:,:])
        for i in range(1,np.shape(A)[0]) :
            s = np.squeeze(A[i,:,:])
            test = np.hstack((test,s))

        U, S, V = np.linalg.svd(test)
        L = max(S)**2

    # List of benchmarked algorithms.
    methods = ['fb', 'fista', 'nesterov']
    
    method = methods[method_i]
    niter = 6000
        
    x_init = np.zeros((n,1))
    if (len(np.shape(A)) == 3) :
        x_init = np.zeros( (np.shape(A)[1], np.shape(A)[2]) )

    x = perform_fb(x_init, 2*L, method, niter, A, y, l, rho)
        
    alpha = x
    return alpha

File: Matlab/forward_backward_optimisation.m
%% Basic Pursuit with Forward-Backward

function x = forward_backward_optimisation(A, y, lambda, method_i, rho)

addpath('optim_function/');

% Matrix and observation.
if nargin < 2
    n = 20;
    p = 1280;
    A = randn(p,64,n);
    A(:,5,[1:50 300:400]) = A(:,5,[1:50 300:400]).*3;
    A(:,30,[100:200 300:400]) = A(:,30,[100:200 300:400]).*2;
    A(:,50,[200:500]) = A(:,5,[200:500]).*1.5+randn;
    y = randn(p,1);
end

% Dimension of the problem.
[p,n] = size(A);

% Regularization parameter.
if nargin < 3
    lambda = 0.2;
end

if nargin < 4
    method_i = 2;
end
if nargin < 5
    rho = 1500;
end

    % Lipschitz constant.
    if length(size(A)) == 2
        L = norm(A);
    elseif length(size(A)) == 3
        % size AA: d2 x d1*d3
        AA = [];
        for i = 1:size(A,1)
            AA = [AA squeeze(A(i,:,:))]; 
        end
        L = max(svd(AA))^2;
    end


% List of benchmarked algorithms.
methods = {'fb', 'fista', 'nesterov'};


% operator callbacks
F = @(x)lambda*norm_L21(x)+rho*norm(x,1);
G = @(x)1/2*norm(y-A*x)^2;
if length(size(A)) == 3
    G = @(x) G3(A, x, y);
end
% Proximal operator of F. 
clear ProxF;
% ProxF = @(x,tau)perform_soft_thresholding(x,lambda*tau); % ProxF = @(x,tau)prox_power(x, lambda*tau, 1);
ProxF = @(x,tau)prox_L21_1(x, lambda*tau, rho*tau);

% Gradient operator of G.
GradG = @(x)A'*(A*x-y);
if length(size(A)) == 3
    GradG = @(x) gradG3(A, x, y);
end

% Function to record the energy.
options.report = @(x)F(x)+G(x);

% Bench the algorithm
options.niter = 6000;

options.method = methods{method_i};
clear x predicted_values;
x_init = zeros(n,1);
if length(size(A)) == 3
    x_init = zeros(size(A,2), size(A,3));
    %ProxF_21 = @(x,tau)prox_L21(x, lambda*tau);
    %x_init = perform_fb(x_init, ProxF_21, GradG, L, options);
end
%disp(['  Constant L before perform fb : ' num2str(L)]);
[x,e] = perform_fb(x_init, ProxF, GradG, 2*L, options); % in toolbox_optim



disp_=0;
if disp_
    nbloc=8;
    blocsize=160;
    % error estimation
    disp(['  **  Correlation between learning and the model, R^2 = ' num2str(R2_train)]);
    clear SStot SSres;
    SStot = sum( (y(:)-mean(y(:))).^2 );
    predicted_values = A*x ;
    SSres = sum( (y(:)-predicted_values(:)).^2 );
    R2_train = 1 - SSres/SStot
    titre = ['learning ' options.method ':  lambda ' num2str(lambda) '. R^2 = ' num2str(R2_train)];
    feature_disp2b(y,blocsize,nbloc); hold on, plot(predicted_values,'LineWidth',2), title(titre);
    
end

end

% for 3D matrices:
function gradG = gradG3(A, x, y)
% A: TxExF
% x: ExF
% y: Tx1
    gradG=zeros(size(x));
    for i=1:size(A,1)
        gradG=gradG+squeeze(A(i,:,:))*(trace(squeeze(A(i,:,:))'*x) - y(i));
    end
end

function G = G3(A, x, y)
    G=0;
    for i=1:size(A,1)
        G=G+(y(i) - trace(squeeze(A(i,:,:))'*x))^2;
    end
end